### PR TITLE
feat: adds count method to resources

### LINKED ISSUE #47: Collections should have `__len__`
Relates to #35 and #46

### PR DESCRIPTION
**Pull request by tdstein**
Resolves #47

### CODE REVIEW

File: src/posit/connect/users.py [hunk 1]
```diff
@@ -151,3 +151,10 @@ def update(self) -> User:
 
     def delete(self) -> None:
         raise NotImplementedError()
+
+    def count(self) -> int:
+        response = self.session.get(self.url)
```
Comment (nealrichardson) [inline src/posit/connect/users.py:156]: Pull less data?

```suggestion
        response = self.session.get(self.url, json={"page_size": 1})
```

File: src/posit/connect/users.py [hunk 2]
```diff
@@ -151,3 +151,10 @@ def update(self) -> User:
...
+        result: dict = response.json()
+        if "total" not in result:
+            raise RuntimeError("'total' not found in response body")
+        return result.get("total", 0)
```
Comment (nealrichardson) [inline src/posit/connect/users.py:160]: Can't see that we'd break this API, and even if we did, I think it's fine to let the `KeyError` bubble up.

```suggestion
        return response.json()["total"]
```

File: tests/posit/connect/__api__/v1/users.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
```
Comment (nealrichardson) [inline tests/posit/connect/__api__/v1/users.json:1]: I'd delete this and just use `tests/posit/connect/__api__/v1/users?page_number=1&page_size=2.json` in the test

File: tests/posit/connect/test_users.py [hunk 4]
```diff
@@ -266,3 +266,13 @@ def test_user_cant_setattr(self):
             match=r"Cannot set attributes: use update\(\) instead",
         ):
             carlos.first_name = "Carlitos"
+
+    @responses.activate
+    def test_count(self):
+        responses.get(
+            "https://connect.example/__api__/v1/users",
+            json=load_mock("v1/users.json"),
+        )
```
Comment (nealrichardson) [inline tests/posit/connect/test_users.py:275]: If you took the suggestion to only request page_size=1 above, add this here:

```
            match=[
                responses.matchers.query_param_matcher(
                    {"page_size": 1}
                )
            ],

File: tinkering.py [hunk 5]
```diff
@@ -1,11 +1,4 @@
...
 with Client() as client:
-    print(client.get("v1/users"))
-    print(client.users.get("f55ca95d-ce52-43ed-b31b-48dc4a07fe13"))
-
-    users = client.users
-    users = users.find(lambda user: user["first_name"].startswith("T"))
-    users = users.find(lambda user: user["last_name"].startswith("S"))
-    user = users.find_one(lambda user: user["user_role"] == "administrator")
-    print(user)
+    print(client.users.count())
```
Comment (nealrichardson) [inline tinkering.py:4]: Do you want to commit this? (Or maybe gitignore the file?)

Comment (tdstein) [reply tinkering.py:4]: It's probably time to delete it. I think we are past the bootstrapping phase

File: src/posit/connect/users.py [hunk 6]
```diff
@@ -148,3 +148,8 @@ def update(self) -> User:
 
     def delete(self) -> None:
         raise NotImplementedError()
+
+    def count(self) -> int:
+        response: requests.Response = self.session.get(self.url, json={"page_size": 1})
```
Comment (nealrichardson) [inline src/posit/connect/users.py:153]: For my own edification: is this type hint necessary? Doesn't it (waves hands) know that `requests.Session.get()` returns `requests.Response`? And that `Response.json()` returns `dict`?

Comment (tdstein) [reply src/posit/connect/users.py:153]: No, it isn't necessary. Once I added the correct type declaration to `session` in the `__init__` method, this type declaration became redundant.

Your understanding is correct 😄

**Pull request merged**